language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractDecoder.java

```diff
@@ -12,11 +12,14 @@
 import java.util.stream.StreamSupport;
 
 public class ContractDecoder {
-  private final Map<Identifier, ? extends ContractCompanion<? extends Contract<?, ?>, ?, ?>>
+  private final Map<
+          Identifier,
+          ? extends ContractCompanion<? extends Contract<?, ?>, ?, ? extends DamlRecord<?>>>
       companions;
 
   public ContractDecoder(
-      Iterable<? extends ContractCompanion<? extends Contract<?, ?>, ?, ?>> companions) {
+      Iterable<? extends ContractCompanion<? extends Contract<?, ?>, ?, ? extends DamlRecord<?>>>
```

Comment:
Here we add a constraint to the fourth type arg to `ContractCompanion`, changing it from `?` to `? extends DamlRecord<?>`. All the Java code-gen classes extend `DamlRecord<?>`, so this is fully compatible with the way we currently use this class. 

The benefit is that we can then know that the `fromJson` of that contract companion will return a type which extends `DamlRecord<?>`, which allows it to be turned back into a `Value` type, which can then be converted into protobuf for use in grpc.